core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,223 @@
+        executorIdToBlacklistExpiryTime.put(exec, expiryTime)
+        executorIdToFailureCount.remove(exec)
+        if (expiryTime < nextExpiryTime) {
+          nextExpiryTime = expiryTime
+        }
+
+        val node = scheduler.getHostForExecutor(exec)
+        val execs = scheduler.getExecutorsAliveOnHost(node).getOrElse(Set())
+        val blacklistedExecs = execs.filter(executorIdToBlacklistExpiryTime.contains(_))
+        if (blacklistedExecs.size >= MAX_FAILED_EXEC_PER_NODE) {
+          logInfo(s"Blacklisting node $node because it has ${blacklistedExecs.size} executors " +
+            s"blacklisted: ${blacklistedExecs}")
+          nodeIdToBlacklistExpiryTime.put(node, expiryTime)
+          // make a copy of the blacklisted nodes so nodeBlacklist() is threadsafe
+          _nodeBlacklist.set(nodeIdToBlacklistExpiryTime.keySet.toSet)
+        }
+      } else {
+        executorIdToFailureCount.put(exec, newTotal)
+      }
+    }
+  }
+
+  def taskSetFailed(stageId: Int): Unit = {
+  }
+
+  def isExecutorBlacklisted(executorId: String): Boolean = {
+    executorIdToBlacklistExpiryTime.contains(executorId)
+  }
+
+  /**
+   * Get the full set of nodes that are blacklisted.  Unlike other methods in this class, this *IS*
+   * thread-safe -- no lock required on a taskScheduler.
+   */
+  def nodeBlacklist(): Set[String] = {
+    _nodeBlacklist.get()
+  }
+
+  def isNodeBlacklisted(node: String): Boolean = {
+    nodeIdToBlacklistExpiryTime.contains(node)
+  }
+
+  def taskSucceeded(
+      stageId: Int,
+      indexInTaskSet: Int,
+      info: TaskInfo,
+      scheduler: TaskSchedulerImpl): Unit = {
+    // no-op intentionally, included just for symmetry.  success to failure ratio is irrelevant, we
+    // just blacklist based on failures.  Furthermore, one success does not clear previous
+    // failures, since the bad node / executor may not fail *every* time
+  }
+
+  def removeExecutor(executorId: String): Unit = {
+    executorIdToBlacklistExpiryTime -= executorId
+    executorIdToFailureCount -= executorId
+  }
+}
+
+
+private[scheduler] object BlacklistTracker extends Logging {
+
+  /**
+   * Return true if the blacklist is enabled, based on the following order of preferences:
+   * 1. Is it specifically enabled or disabled?
+   * 2. Is it enabled via the legacy timeout conf?
+   * 3. Use the default for the spark-master:
+   *   - off for local mode
+   *   - on for distributed modes (including local-cluster)
+   */
+  def isBlacklistEnabled(conf: SparkConf): Boolean = {
+    conf.get(config.BLACKLIST_ENABLED) match {
+      case Some(isEnabled) =>
+        isEnabled
+      case None =>
+      // if they've got a non-zero setting for the legacy conf, always enable the blacklist,
+      // otherwise, use the default based on the cluster-mode (off for local-mode, on otherwise).
+        val legacyKey = config.BLACKLIST_LEGACY_TIMEOUT_CONF.key
+      conf.getOption(legacyKey) match {
```

Comment:
Indentation. Also, this can be:

```
conf.getOption(config.BLACKLIST_LEGACY_TIMEOUT_CONF) match {
  case Some(timeout) if timeout > 0 => ...
  case Some(timeout) => ...
  case None => ...
}
```